python/ffsim/states/slater.py

```diff
@@ -0,0 +1,271 @@
+        else:
+            strings_a = _sample_spinless_direct(rdm_a, n_a, shots, rng)
+
+        if n_b == 0:
+            strings_b = [0 for i in range(shots)]
+        elif n_b == norb:
+            strings_b = [sum(1 << j for j in range(norb)) for i in range(shots)]
+        else:
+            strings_b = _sample_spinless_direct(rdm_b, n_b, shots, rng)
+
+        strings_a = restrict_bitstrings(
+            strings_a, orbs_a, bitstring_type=BitstringType.INT
+        )
+        strings_b = restrict_bitstrings(
+            strings_b, orbs_b, bitstring_type=BitstringType.INT
+        )
+
+        if concatenate:
+            strings = concatenate_bitstrings(
+                strings_a,
+                strings_b,
+                BitstringType.INT,
+                length=len(orbs_a),
+            )
+            return convert_bitstring_type(
+                strings,
+                BitstringType.INT,
+                bitstring_type,
+                length=len(orbs_a) + len(orbs_b),
+            )
+
+        return convert_bitstring_type(
+            strings_a,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs_a),
+        ), convert_bitstring_type(
+            strings_b,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs_b),
+        )
+
+
+def _generate_conditionals_unormalized(
+    rdm: np.ndarray, sample: list[int], empty_orbitals: set[int], marginal: float
+) -> tuple[np.ndarray, np.ndarray]:
+    """Generates the conditional and marginal probabilities for adding a particle.
+
+    This is a step of the autoregressive sampling, and uses Bayes's rule.
+
+    Args:
+        rdm: A Numpy array with the one-body reduced density matrix.
+        pos_array: A numpy array with the positions of the particles.
+        empty_orbitals: A numpy array with the empty orbitals that a new particle
+            may occupy. The sorted union of ``pos_array`` and ``empty_orbitals``
+            must be equal to ``numpy.arange(num_orbitals)``.
+        marginal: The marginal probability associated to having the particles
+            in the positions determined by ``pos_array``.
+    Returns:
+        A tuple of Numpy arrays. The first is the unormalized conditionals for
+        adding a particle to any of the empty orbitals specified in the input.
+        The second is the marginal corresponding to having the particles in the
+        position array and one extra in all possible empty orbitals. Both arrays
+        follow the same order as ``empty_orbitals``.
+
+    """
+    conditionals = np.zeros(len(empty_orbitals), dtype=complex)
+    marginals = np.zeros(len(empty_orbitals), dtype=complex)
+
+    for i, orbital in enumerate(empty_orbitals):
+        new_sample = copy.deepcopy(sample)
+        new_sample.append(orbital)
+        rest_rdm = rdm[new_sample, :]
+        rest_rdm = rest_rdm[:, new_sample]
+        marginals[i] = np.linalg.det(rest_rdm)
```

Comment:
I just did that. It complains that we are casting a complex number into a real one. This happens because the RDM is allowed ot be complex, however since it is a Hermitian matrix, its determinant is guaranteed to be real.